torch/csrc/perf_tools/ThroughputBenchmark-inl.h

```diff
@@ -0,0 +1,143 @@
+#pragma once
+
+#include <random>
+#include <thread>
+
+#include <torch/csrc/utils/pybind.h>
+#include <torch/csrc/jit/pybind_utils.h>
+
+namespace torch {
+namespace throughput_benchmark {
+namespace detail {
+
+template <class Input, class Output, class Model>
+BenchmarkExecutionStats BenchmarkHelper<Input, Output, Model>::benchmark(
+    const BenchmarkConfig& config) const {
+  CHECK(initialized_);
+  TORCH_CHECK(
+      config.num_worker_threads == 1,
+      "Only parallelization by callers is supported");
+
+  // We pre-generate inputs for each of the threads and avoid threadsafety
+  // issues for random calls. As thread safe random is not provided in the
+  // standard library. We also replicate IValues. So we don't need to copy
+  // them during the main benchmark phase.
+  std::vector<std::vector<Input>> thread_inputs(config.num_calling_threads);
+  std::vector<size_t> input_iters(config.num_calling_threads);
+  {
+    std::random_device seeder;
+    std::mt19937 engine(seeder());
+    TORCH_CHECK(!inputs_.empty(), "Please provide benchmark inptus");
+    std::uniform_int_distribution<int> dist(0, inputs_.size() - 1);
+
+    for (int thread_id = 0; thread_id < config.num_calling_threads;
+         ++thread_id) {
+      // Just in case we generate num_iters inputs for each of the threads
+      // This was if one thread does all the work we will be fine
+      for (int i = 0; i < config.num_iters + config.num_warmup_iters; ++i) {
+        thread_inputs[thread_id].push_back(inputs_[dist(engine)]);
+      }
+      input_iters[thread_id] = 0;
+    }
+  }
+
+  std::mutex m;
+  std::condition_variable cv;
+  // TODO: add GUARDED_BY once it is available
+  int64_t initialized{0};
+  int64_t finished{0};
+  bool start{false};
+  std::atomic<int64_t> num_forwards{0};
+  std::vector<std::thread> callers;
+
+  // Children need the GIL
+  AutoNoGIL no_gil_guard;
+  for (auto thread_id = 0; thread_id < config.num_calling_threads;
+       ++thread_id) {
+    callers.emplace_back([&, thread_id]() {
+      // We use conditional variable as a barrier to make sure each thread
+      // performs required warmeup iterations before we start measuring
+      for (auto j = 0; j < config.num_warmup_iters; ++j) {
+        if (std::is_same<Input, ModuleInput>::value) {
+          AutoGIL gil_guard;
```

Comment:
@dzhulgakov , I was not able to get this code working even with custom move constructor injecting GIL :'( So ended up just GILing here. I will try mutable reference, it will most likely work, thanks for the idea!

Reply:
How does it fail if one std::move() for py::object? I find it surprising

Reply:
OMG, all this time it was destructor of the result, not the move itself! So my very first version I did here got the GIL right, I just should have not returned anything here.